### ISSUE TITLE
[explicit-module-boundary-types] return annotation required although it immediately returns another function

### ISSUE THREAD

**Issue by cbn-falias** [typescript-eslint/typescript-eslint#4920]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**
```JSON
{
  "rules": {
    "@typescript-eslint/explicit-module-boundary-types": [
      "warn",
      {
        "allowHigherOrderFunctions": true
      }
    ]
  }
}
```

```TS
const moduleFn = (prefix: string, id: string): string => {
  return prefix + id;
};

export const customFn = (id: string) => moduleFn('Foo', id);
```

**Expected Result**
No warning for function `customFn`.

**Actual Result**
`Missing return type on function.eslint[@typescript-eslint/explicit-module-boundary-types](https://typescript-eslint.io/rules/explicit-module-boundary-types)`

**Additional Info**
I'm not sure if this rule/option should actually cover those use cases but it would be quite helpful.
My "real life use cases" are that we have some wrapper functions for React packages and it's quite cumbersome sometimes to find the correct return type.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.22.1-alpha.8` |
| `@typescript-eslint/parser`        | `X.Y.Z` |
| `TypeScript`                       | `4.1.2` |
| `ESLint`                           | `8.14.0` |
| `node`                             | `16.14.2` |

**Comment by bradzacher** [comment]
There is no option that covers this.

A function that returns a function call. Is just a function that returns a value. I don't see any reason why this case should be treated any differently than any other return value.

Sometimes the return type can be non-trivial to resolve exactly - but this is part of the reason this rule exists - to help you write out clear contracts for your boundary to make the API clear and more reviewable.
